--- a/arch/mips/bmips/setup.c
+++ b/arch/mips/bmips/setup.c
@@ -107,6 +107,7 @@ static void bcm6368_quirks(void)
 
 static const struct bmips_quirk bmips_quirk_list[] = {
 	{ "brcm,bcm3368",		&bcm3368_quirks			},
+	{ "brcm,bcm3380",		&bcm6368_quirks			},
 	{ "brcm,bcm3384-viper",		&bcm3384_viper_quirks		},
 	{ "brcm,bcm33843-viper",	&bcm3384_viper_quirks		},
 	{ "brcm,bcm6328",		&bcm6328_quirks			},
